diff -Naur linux-libc-headers-2.6.16.16/include/asm-arm/unistd.h linux-libc-headers-2.6.16.16-stm/include/asm-arm/unistd.h
--- linux-libc-headers-2.6.16.16/include/asm-arm/unistd.h	2006-05-14 22:52:37.000000000 +0100
+++ linux-libc-headers-2.6.16.16-stm/include/asm-arm/unistd.h	2006-07-21 09:09:31.723220696 +0100
@@ -17,7 +17,7 @@
 
 #define __NR_OABI_SYSCALL_BASE	0x900000
 
-#if defined(__thumb__) || defined(__ARM_EABI__)
+#if 1 /* defined(__thumb__) || defined(__ARM_EABI__) */
 #define __NR_SYSCALL_BASE	0
 #else
 #define __NR_SYSCALL_BASE	__NR_OABI_SYSCALL_BASE
@@ -309,7 +309,7 @@
 #define __NR_mq_getsetattr		(__NR_SYSCALL_BASE+279)
 #define __NR_waitid			(__NR_SYSCALL_BASE+280)
 
-#if defined(__ARM_EABI__)  /* reserve these for un-muxing socketcall */
+#if 1 /* defined(__ARM_EABI__) */  /* reserve these for un-muxing socketcall */
 #define __NR_socket			(__NR_SYSCALL_BASE+281)
 #define __NR_bind			(__NR_SYSCALL_BASE+282)
 #define __NR_connect			(__NR_SYSCALL_BASE+283)
@@ -329,7 +329,7 @@
 #define __NR_recvmsg			(__NR_SYSCALL_BASE+297)
 #endif
 
-#if defined(__ARM_EABI__)  /* reserve these for un-muxing ipc */
+#if 1 /* defined(__ARM_EABI__) */  /* reserve these for un-muxing ipc */
 #define __NR_semop			(__NR_SYSCALL_BASE+298)
 #define __NR_semget			(__NR_SYSCALL_BASE+299)
 #define __NR_semctl			(__NR_SYSCALL_BASE+300)
@@ -347,7 +347,7 @@
 #define __NR_request_key		(__NR_SYSCALL_BASE+310)
 #define __NR_keyctl			(__NR_SYSCALL_BASE+311)
 
-#if defined(__ARM_EABI__)  /* reserved for un-muxing ipc */
+#if 1 /* defined(__ARM_EABI__) */  /* reserved for un-muxing ipc */
 #define __NR_semtimedop			(__NR_SYSCALL_BASE+312)
 #endif
 
@@ -375,7 +375,7 @@
 #define __sys1(x) __sys2(x)
 
 #ifndef __syscall
-#if defined(__thumb__) || defined(__ARM_EABI__)
+#if 1 /* defined(__thumb__) || defined(__ARM_EABI__) */
 #define __SYS_REG(name) register long __sysreg __asm__("r7") = __NR_##name;
 #define __SYS_REG_LIST(regs...) "r" (__sysreg) , ##regs
 #define __syscall(name) "swi\t0"
