--- libc/sysdeps/gnu/Makefile.orig	2005-11-29 09:55:12.000000000 +0000
+++ libc/sysdeps/gnu/Makefile	2005-11-29 10:06:04.000000000 +0000
@@ -42,7 +42,7 @@
 endif
 	$(AWK) -v maxerr=`\
 	  $(CC) -S $(CPPFLAGS) $(CFLAGS) -DNOT_IN_libc -DEMIT_ERR_MAX $< -o - \
-	  | sed -n 's/^.*@@@[^0-9]*\([0-9]*\)[^0-9]*@@@.*$$/\1/p'` \
+	  | sed '/^##/d' | sed -n 's/^.*@@@[^0-9]*\([0-9]*\)[^0-9]*@@@.*$$/\1/p'` \
 	       -f $(..)sysdeps/gnu/errlist-compat.awk \
 	       $(wildcard $(patsubst %,$(..)%/Versions,\
 		 $(config-sysdirs) $(add-ons))) > $@T
