--- libc/./sysdeps/unix/sysv/linux/st200/mmap.c.orig	2006-05-23 16:39:13.000000000 +0100
+++ libc/./sysdeps/unix/sysv/linux/st200/mmap.c	2006-05-23 16:39:24.000000000 +0100
@@ -34,7 +34,7 @@
     return MAP_FAILED;
   }
 
-  return (void *) INLINE_SYSCALL (mmap, 6, addr, len, prot, flags, fd,
+  return (void *) INLINE_SYSCALL (mmap2, 6, addr, len, prot, flags, fd,
 				  offset >> MMAP2_PAGE_SHIFT);
 }
 
