Index: gdb-6.5/gdb/testsuite/gdb.asm/sh.inc
===================================================================
--- gdb-6.5.orig/gdb/testsuite/gdb.asm/sh.inc	2005-03-07 21:16:50.000000000 +0000
+++ gdb-6.5/gdb/testsuite/gdb.asm/sh.inc	2007-09-18 17:03:03.000000000 +0100
@@ -65,7 +65,8 @@
 
 	comment "crt0 startup"
 	.macro gdbasm_startup
-	mov.l	.stackaddr,r15
+	comment "mov.l	.stackaddr,r15"
+	nop
 	bra	.afterstackaddr
 	nop
 	nop
Index: gdb-6.5/gdb/testsuite/gdb.base/sigall.c
===================================================================
--- gdb-6.5.orig/gdb/testsuite/gdb.base/sigall.c	1999-06-28 17:04:02.000000000 +0100
+++ gdb-6.5/gdb/testsuite/gdb.base/sigall.c	2007-09-18 17:03:03.000000000 +0100
@@ -1,10 +1,6 @@
 #include <signal.h>
 #include <unistd.h>
 
-#ifdef __sh__
-#define signal(a,b)	/* Signals not supported on this target - make them go away */
-#endif
-
 /* Signal handlers, we set breakpoints in them to make sure that the
    signals really get delivered.  */
 
