Index: gcc/gcc/config/sh/sh.c
===================================================================
--- gcc.orig/gcc/config/sh/sh.c	2008-05-07 13:46:36.000000000 +0100
+++ gcc/gcc/config/sh/sh.c	2008-05-07 13:47:55.000000000 +0100
@@ -4108,6 +4108,15 @@
 	       && ! TARGET_SMALLCODE)
 	new_align = 4;
 
+      /* There is a possibility that a bf is transformed into a bf/s by the
+	 delay slot scheduler, Conservatively handle this case.  */
+      if (GET_CODE (from) == JUMP_INSN
+	  /* Don't mess with ADDR_DIFF_VEC */
+	  && GET_CODE (PATTERN (from)) == SET
+	  && get_attr_type (from) == TYPE_CBRANCH)
+	inc += 2;
+
+
       if (found_si)
 	{
 	  count_si += inc;
Index: gcc/gcc/ChangeLog.STM
===================================================================
--- gcc.orig/gcc/ChangeLog.STM	2008-05-07 13:49:07.000000000 +0100
+++ gcc/gcc/ChangeLog.STM	2008-05-07 13:49:27.000000000 +0100
@@ -1,3 +1,7 @@
+2008-05-07  Christian Bruel  <christian.bruel@st.com>
+
+       https://bugzilla.stlinux.com/show_bug.cgi?id=3891
+       * config/sh/sh.c (find_barrier): Increase size of conditional branches.
 
 2008-01-28  Christian Bruel  <christian.bruel@st.com>
 
