Index: gcc-4.2.4/gcc/Makefile.in
===================================================================
--- gcc-4.2.4.orig/gcc/Makefile.in	2008-10-09 11:01:19.000000000 +0100
+++ gcc-4.2.4/gcc/Makefile.in	2008-10-09 11:29:21.000000000 +0100
@@ -2668,7 +2668,7 @@
 # FIXME: writing proper dependencies for this is a *LOT* of work.
 libbackend.o : $(OBJS-common:.o=.c) $(out_file) \
   insn-config.h insn-flags.h insn-codes.h insn-constants.h \
-  insn-attr.h  $(DATESTAMP) $(BASEVER) $(DEVPHASE)
+  insn-attr.h  $(DATESTAMP) $(BASEVER) $(DEVPHASE) gcov-iov.h
 	$(CC) $(ALL_CFLAGS) $(ALL_CPPFLAGS) \
 	  -DTARGET_NAME=\"$(target_noncanonical)\" \
 	  -DLOCALEDIR=\"$(localedir)\" \
