diff -urN gdb-6.8.old/gdb/doc/Makefile.in gdb-6.8/gdb/doc/Makefile.in
--- gdb-6.8.old/gdb/doc/Makefile.in	2008-01-01 22:53:14.000000000 +0000
+++ gdb-6.8/gdb/doc/Makefile.in	2009-08-03 15:57:51.000000000 +0000
@@ -331,8 +331,7 @@
 
 # GDB MANUAL: info file
 gdb.info: ${GDB_DOC_FILES}
-	$(MAKEINFO) -I ${READLINE_DIR} -I ${GDBMI_DIR} -I $(srcdir) \
-		-o gdb.info $(srcdir)/gdb.texinfo
+	touch gdb.info
 
 # GDB MANUAL: roff translations
 # Try to use a recent texi2roff.  v2 was put on prep in jan91.
