diff -r -N -u opt.bkup/STM/STLinux-2.3/devkit/sh4/target/usr/lib/libstdc++.la opt/STM/STLinux-2.3/devkit/sh4/target/usr/lib/libstdc++.la
--- opt.bkup/STM/STLinux-2.3/devkit/sh4/target/usr/lib/libstdc++.la	2010-08-10 09:30:46.910809737 +0000
+++ opt/STM/STLinux-2.3/devkit/sh4/target/usr/lib/libstdc++.la	2010-08-10 09:31:49.074814755 +0000
@@ -32,4 +32,4 @@
 dlpreopen=''
 
 # Directory that this library needs to be installed in:
-libdir='/usr/lib'
+libdir="/opt/STM/STLinux-2.3/devkit/sh4/target/usr/lib"
diff -r -N -u opt.bkup/STM/STLinux-2.3/devkit/sh4/target/usr/lib/libsupc++.la opt/STM/STLinux-2.3/devkit/sh4/target/usr/lib/libsupc++.la
--- opt.bkup/STM/STLinux-2.3/devkit/sh4/target/usr/lib/libsupc++.la	2010-08-10 09:30:46.910809737 +0000
+++ opt/STM/STLinux-2.3/devkit/sh4/target/usr/lib/libsupc++.la	2010-08-10 09:32:03.214309570 +0000
@@ -32,4 +32,4 @@
 dlpreopen=''
 
 # Directory that this library needs to be installed in:
-libdir='/usr/lib'
+libdir="/opt/STM/STLinux-2.3/devkit/sh4/target/usr/lib"
