Fix watchpoint removal.

Signed-off-by: Chris Smith <chris.smith@st.com>

Index: gdb-6.8/gdb/target.c
===================================================================
--- gdb-6.8.orig/gdb/target.c	2008-09-11 10:35:35.000000000 +0100
+++ gdb-6.8/gdb/target.c	2008-09-11 14:38:00.000000000 +0100
@@ -2389,10 +2389,10 @@
 {
   int retval;
 
-  retval = debug_target.to_insert_watchpoint (addr, len, type);
+  retval = debug_target.to_remove_watchpoint (addr, len, type);
 
   fprintf_unfiltered (gdb_stdlog,
-		      "target_insert_watchpoint (0x%lx, %d, %d) = %ld\n",
+		      "target_remove_watchpoint (0x%lx, %d, %d) = %ld\n",
 		      (unsigned long) addr, len, type, (unsigned long) retval);
   return retval;
 }
