--- gdb-6.8_a/gdb/configure.ac	2009-05-20 19:19:20.000000000 +0530
+++ gdb-6.8/gdb/configure.ac	2009-05-20 19:29:14.000000000 +0530
@@ -1442,7 +1442,7 @@
 # ST Microelectronics add-on - need stdc++ for Linux
 echo $ECHO_N "checking ST Microelectronics add-on... $ECHO_C" >&6
 case ${host} in
-    i[[3456]]86-*-linux*)
+    i[[3456]]86-*-linux* | *x86_64*)
     echo "${ECHO_T}add -lstdc++ -ldl -lpthread" >&6
     LIBS="-lstdc++ -ldl -lpthread $LIBS"
     ;;
