--- linux-st231-2.6.11.12_stm22-19.orig/include/asm-st200/unistd.h
+++ linux-st231-2.6.11.12_stm22-19/include/asm-st200/unistd.h
@@ -15,7 +15,7 @@
 
 #define __NR_syscall                0
 #define __NR_exit                   1
-#define __NR_fork                   2
+/*#define __NR_fork                   2*/
 #define __NR_read                   3
 #define __NR_write                  4
 #define __NR_open                   5
@@ -42,7 +42,7 @@
 #define __NR_ptrace                26
 /* #define __NR_alarm                 27 */
 #define __NR_fstat                 28
-#define __NR_pause                 29
+/*#define __NR_pause                 29*/
 #define __NR_utimes                30
 #define __NR_connect               31
 #define __NR_listen                32
@@ -78,7 +78,7 @@
 #define __NR_ustat                 62
 #define __NR_dup2                  63
 #define __NR_getppid               64
-#define __NR_getpgrp               65
+/*#define __NR_getpgrp               65*/
 #define __NR_setsid                66
 #define __NR_pivot_root            67
 /* #define __NR_sgetmask              68 */
@@ -103,7 +103,7 @@
 #define __NR_swapon                87
 #define __NR_reboot                88
 /* #define __NR_mmap2                 89 */
-#define __NR_mmap                  90
+#define __NR_mmap2                 90
 #define __NR_munmap                91
 #define __NR_truncate              92
 #define __NR_ftruncate             93
@@ -126,7 +126,7 @@
 #define __NR_getcwd               110
 #define __NR_vhangup              111
 #define __NR_fstat64              112
-#define __NR_vfork                113
+/*#define __NR_vfork                113*/
 #define __NR_wait4                114
 #define __NR_swapoff              115
 #define __NR_sysinfo              116
@@ -140,17 +140,17 @@
 #define __NR_adjtimex             124
 #define __NR_mprotect             125
 /* #define __NR_sigprocmask          126 */
-#define __NR_create_module        127
+/*#define __NR_create_module        127*/
 #define __NR_init_module          128
 #define __NR_delete_module        129
-#define __NR_get_kernel_syms      130
+/*#define __NR_get_kernel_syms      130*/
 #define __NR_quotactl             131
 #define __NR_getpgid              132
 #define __NR_fchdir               133
 #define __NR_bdflush              134
 #define __NR_sysfs                135
 #define __NR_personality          136
-#define __NR_afs_syscall          137 /* Syscall for Andrew File System */
+/*#define __NR_afs_syscall          137  Syscall for Andrew File System */
 #define __NR_setfsuid             138
 #define __NR_setfsgid             139
 #define __NR__llseek              140
@@ -451,11 +451,8 @@ static __inline__ _syscall3(int,execve,c
 
 
 #ifdef __KERNEL__
-/* FIXME : We should ditch pause and getpgrp before we release */
 
 #define __ARCH_WANT_STAT64
-#define __ARCH_WANT_SYS_PAUSE
-#define __ARCH_WANT_SYS_GETPGRP
 #define __ARCH_WANT_SYS_LLSEEK
 #define __ARCH_WANT_SYS_RT_SIGACTION
 
Index: linux-st231-2.6.11.12_stm22-19/arch/st200/kernel/syscall.S
===================================================================
