Index: gcc-4.2.4/libstdc++-v3/configure
===================================================================
--- gcc-4.2.4.orig/libstdc++-v3/configure	2008-10-09 11:04:54.000000000 +0100
+++ gcc-4.2.4/libstdc++-v3/configure	2008-10-09 11:20:21.000000000 +0100
@@ -7514,6 +7514,9 @@
 cat >>conftest.$ac_ext <<_ACEOF
 /* end confdefs.h.  */
 #include <complex.h>
+#ifdef __UCLIBC__
+#error ugly hack to make sure configure test fails here for cross until uClibc supports the complex funcs
+#endif
 int
 main ()
 {
