Index: gcc/gcc/config.gcc
===================================================================
--- gcc.orig/gcc/config.gcc	2008-07-18 11:11:06.000000000 +0100
+++ gcc/gcc/config.gcc	2008-07-18 11:12:16.000000000 +0100
@@ -1968,7 +1968,7 @@
         ;;
 sh-*-elf* | sh[12346l]*-*-elf* | sh*-*-kaos* | \
 sh-*-symbianelf* | sh[12346l]*-*-symbianelf* | \
-  sh-*-linux* | sh[346lbe]*-*-linux* | \
+  sh*-*-linux* | sh[346lbe]*-*-linux* | \
   sh-*-netbsdelf* | shl*-*-netbsdelf* | sh5-*-netbsd* | sh5l*-*-netbsd* | \
    sh64-*-netbsd* | sh64l*-*-netbsd*)
 	tmake_file="${tmake_file} sh/t-sh sh/t-elf"
Index: gcc/libgomp/configure
===================================================================
--- gcc.orig/libgomp/configure	2008-02-01 17:06:50.000000000 +0000
+++ gcc/libgomp/configure	2008-07-18 11:12:16.000000000 +0100
@@ -3893,7 +3893,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/gcc/config/cris/linux.h
===================================================================
--- gcc.orig/gcc/config/cris/linux.h	2007-09-01 16:28:30.000000000 +0100
+++ gcc/gcc/config/cris/linux.h	2008-07-18 11:12:16.000000000 +0100
@@ -73,7 +73,11 @@
 #define CRIS_DEFAULT_CPU_VERSION CRIS_CPU_NG
 
 #undef CRIS_SUBTARGET_VERSION
-#define CRIS_SUBTARGET_VERSION " - cris-axis-linux-gnu"
+#if UCLIBC_DEFAULT
+# define CRIS_SUBTARGET_VERSION " - cris-axis-linux-uclibc"
+#else
+# define CRIS_SUBTARGET_VERSION " - cris-axis-linux-gnu"
+#endif
 
 #define GLIBC_DYNAMIC_LINKER "/lib/ld.so.1"
 
Index: gcc/libstdc++-v3/configure
===================================================================
--- gcc.orig/libstdc++-v3/configure	2007-06-29 00:02:05.000000000 +0100
+++ gcc/libstdc++-v3/configure	2008-07-18 11:12:16.000000000 +0100
@@ -4283,7 +4283,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/zlib/configure
===================================================================
--- gcc.orig/zlib/configure	2007-01-17 18:10:26.000000000 +0000
+++ gcc/zlib/configure	2008-07-18 11:12:16.000000000 +0100
@@ -3429,7 +3429,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/libobjc/configure
===================================================================
--- gcc.orig/libobjc/configure	2006-10-15 08:42:57.000000000 +0100
+++ gcc/libobjc/configure	2008-07-18 11:12:16.000000000 +0100
@@ -3314,7 +3314,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/libgfortran/configure
===================================================================
--- gcc.orig/libgfortran/configure	2008-02-01 17:06:49.000000000 +0000
+++ gcc/libgfortran/configure	2008-07-18 11:12:16.000000000 +0100
@@ -3721,7 +3721,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/libmudflap/configure
===================================================================
--- gcc.orig/libmudflap/configure	2006-12-04 11:13:07.000000000 +0000
+++ gcc/libmudflap/configure	2008-07-18 11:12:16.000000000 +0100
@@ -5394,7 +5394,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/boehm-gc/configure
===================================================================
--- gcc.orig/boehm-gc/configure	2008-02-01 17:06:49.000000000 +0000
+++ gcc/boehm-gc/configure	2008-07-18 11:12:16.000000000 +0100
@@ -4323,7 +4323,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/libffi/configure
===================================================================
--- gcc.orig/libffi/configure	2008-02-01 17:06:49.000000000 +0000
+++ gcc/libffi/configure	2008-07-18 11:12:16.000000000 +0100
@@ -3460,7 +3460,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/libssp/configure
===================================================================
--- gcc.orig/libssp/configure	2006-10-15 08:42:57.000000000 +0100
+++ gcc/libssp/configure	2008-07-18 11:12:16.000000000 +0100
@@ -4480,7 +4480,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/contrib/regression/objs-gcc.sh
===================================================================
--- gcc.orig/contrib/regression/objs-gcc.sh	2005-08-15 01:41:31.000000000 +0100
+++ gcc/contrib/regression/objs-gcc.sh	2008-07-18 11:12:16.000000000 +0100
@@ -105,6 +105,10 @@
  then
   make all-gdb all-dejagnu all-ld || exit 1
   make install-gdb install-dejagnu install-ld || exit 1
+elif [ $H_REAL_TARGET = $H_REAL_HOST -a $H_REAL_TARGET = i686-pc-linux-uclibc ]
+ then
+  make all-gdb all-dejagnu all-ld || exit 1
+  make install-gdb install-dejagnu install-ld || exit 1
 elif [ $H_REAL_TARGET = $H_REAL_HOST ] ; then
   make bootstrap || exit 1
   make install || exit 1
Index: gcc/libjava/classpath/ltconfig
===================================================================
--- gcc.orig/libjava/classpath/ltconfig	2006-08-07 21:37:50.000000000 +0100
+++ gcc/libjava/classpath/ltconfig	2008-07-18 11:12:16.000000000 +0100
@@ -603,7 +603,7 @@
 
 # Transform linux* to *-*-linux-gnu*, to support old configure scripts.
 case $host_os in
-linux-gnu*) ;;
+linux-gnu*|linux-uclibc*) ;;
 linux*) host=`echo $host | sed 's/^\(.*-.*-linux\)\(.*\)$/\1-gnu\2/'`
 esac
 
@@ -1251,7 +1251,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   version_type=linux
   need_lib_prefix=no
   need_version=no
Index: gcc/libjava/classpath/configure
===================================================================
--- gcc.orig/libjava/classpath/configure	2007-01-17 18:10:26.000000000 +0000
+++ gcc/libjava/classpath/configure	2008-07-18 11:12:16.000000000 +0100
@@ -5307,7 +5307,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/libjava/configure
===================================================================
--- gcc.orig/libjava/configure	2008-02-01 17:06:50.000000000 +0000
+++ gcc/libjava/configure	2008-07-18 11:12:16.000000000 +0100
@@ -5424,7 +5424,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/libtool.m4
===================================================================
--- gcc.orig/libtool.m4	2008-07-18 10:37:55.000000000 +0100
+++ gcc/libtool.m4	2008-07-18 11:12:16.000000000 +0100
@@ -743,7 +743,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   lt_cv_deplibs_check_method=pass_all
   ;;
 
Index: gcc/ltconfig
===================================================================
--- gcc.orig/ltconfig	2008-07-18 10:37:55.000000000 +0100
+++ gcc/ltconfig	2008-07-18 11:12:16.000000000 +0100
@@ -603,7 +603,7 @@
 
 # Transform linux* to *-*-linux-gnu*, to support old configure scripts.
 case $host_os in
-linux-gnu*) ;;
+linux-gnu*|linux-uclibc*) ;;
 linux*) host=`echo $host | sed 's/^\(.*-.*-linux\)\(.*\)$/\1-gnu\2/'`
 esac
 
@@ -1252,7 +1252,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux*)
   version_type=linux
   need_lib_prefix=no
   need_version=no
